src/hotspot/share/services/management.cpp

```diff
@@ -2095,7 +2095,21 @@ jlong Management::ticks_to_ms(jlong ticks) {
   return (jlong)(((double)ticks / (double)os::elapsed_frequency())
                  * (double)1000.0);
 }
-#endif // INCLUDE_MANAGEMENT
+
+// Gets the amount of memory allocated on the Java heap since JVM launch.
+JVM_ENTRY(jlong, jmm_GetAllThreadAllocatedMemory(JNIEnv *env))
+    // The result may be "too small" because the threads iterator takes a
+    // snapshot and new threads may be created after result is initialized.
+    // There's no check for terminated threads in the loop because their
+    // allocated byte counts aren't included in exited_allocated_bytes
+    // when result is initialized.
+    jlong result = ThreadService::exited_allocated_bytes();
+    for (JavaThreadIteratorWithHandle jtiwh; JavaThread *thread = jtiwh.next();) {
```

Comment:
If you call `exited_allocated_bytes` whilst you have an active `ThreadsListHandle` then you at least ensure you don't miss accounting for threads that are just about to terminate.

Reply:
Do you mean something like:
```cpp
    JavaThreadIteratorWithHandle jtiwh;
    jlong result = ThreadService::exited_allocated_bytes();
    while (JavaThread *thread = jtiwh.next()) {
      ...
```
That would be fine for me. Otherwise I agree with the current compromise between accuracy and speed.

Reply:
Yes. Though I realize my comment needed to make clear that you need a thread to update the count after it has removed itself from the thread-list, for this to be an improvement.

Reply:
Done as Volker suggests.